lisa/datautils.py

```diff
@@ -416,6 +416,85 @@ def merge(df1, df2):
     return functools.reduce(merge, df_list)
 
 
+def df_delta(pre_df, post_df, on, post_col=None):
+    """
+    Add a ``delta`` column to the `pre_df`` dataframe containing the duration
+    before a consecutive event from ``post_df`` is emitted.
+
+    Rows from ``pre_df`` and ``post_df`` are grouped by the ``on`` columns.
+    E.g.: This would be ['pid', 'comm'] to group by task.
+
+    This is usefull to concatenate two dataframe containing consecutive events.
+    A typical usecase would be adding pre/post events at the entrance/exit of a
+    function.
+
+    If a 'pre' event has no corresponding 'post' event, the 'pre' event is
+    dropped.
+
+    :param pre_df: Dataframe containing the events that start a record. All
+        the columns of this dataframe are available in the returned dataframe.
+    :type df: pandas.DataFrame
+
+    :param post_df: Dataframe containing the events that stop a record. To add
+        columns from this dataframe to the returned dataframe, use
+        ``post_col``.
+    :type df: pandas.DataFrame
+
+    :param on: Columns used to group ``pre_df`` and ``post_df``.
+        E.g.: This would be ['pid', 'comm'] to group by task.
+    :type on: list(str)
+
+    :param post_col: Columns to take from the ``post_df`` and add to the
+        returned dataframe.
+    :type post_col: list(str)
+
+    :returns: a :class:`pandas.DataFrame` with:
+
+        * All the columns from the ``pre_df`` dataframe.
+        * A ``delta`` column (duration between the emission of the ``pre_df``
+            and its consecutive ``post_df``).
+        * All the ``post_col`` columns from the ``post_df``.
+    """
+    pre_df = pre_df.copy(deep=False)
+    post_df = post_df.copy(deep=False)
+
+    # Only keep useful columns in post_df.
+    post_df = post_df[on + post_col]
+
+    # Merge on columns common to the two dfs to avoid overlapping of names.
+    on_col = list(set(pre_df).intersection(set(post_df)))
+
+    # Merging on nullable types convert columns to object.
+    # Merging on non-nullable types converts integer/boolean to float.
+    # Thus, let the on_col non-nullable and converts the others to nullable.
+    pre_df_col = list(set(pre_df) - set(on_col))
+    post_df_col = list(set(post_df) - set(on_col))
+    pre_df[pre_df_col] = df_convert_to_nullable(pre_df[pre_df_col])
+    post_df[post_df_col] = df_convert_to_nullable(post_df[post_df_col])
+
+    # Merge. Don't allow column renaming.
+    df = pd.merge(pre_df, post_df, left_index=True, right_index=True, on=on_col,
+                  how='outer', suffixes=(False, False))
+    df.index.name = 'Time'
+    df.reset_index(inplace=True)
+
+    # Create 'delta' column + additionnal post_col.
+    grouped = df.groupby(on, observed=True, sort=False)
+    new_columns = dict(
+        delta=grouped['Time'].transform(lambda time: time.diff().shift(-1)),
```

Comment:
This assumes that:
* Events are perfectly interleaved, which might not be the case. If for some reason we end up with a "pre->post->post->pre" sequence, the delta will become silently meaningless.
* The dataframe starts with a "pre" event, which is not enforced. If it stats with a "post" event, we will end up measuring the delta between a post and the next pre, which is not what we want.

You could add an "is_pre" column that would be True if it's a pre-event and False otherwise:
```
pre_df['is_pre'] = True
post_df['is_pre'] = False
```
And the on each group:
```
# Only keep the rows for which the "is_pre" status is different from the previous one
df = df.loc[df['is_pre'] != df['is_pre'].shift()]
# Ignore the first row if it's a "post" event
if len(df) >= 1 and not df['is_pre'].iat[0]:
     df = df.iloc[1:]
```
You will have to use the iterable interface of "grouped" that yields `(group, subdf)` tuples with `subdf` being a subset of the initial dataframe, so you can apply the processing using multiple columns on that `subdf` (rather than one processing on the `Time` column directly on the `grouped` object).

Reply:
> You will have to use the iterable interface of "grouped" that yields (group, subdf) tuples

If we process each group individually to remove the isolated pre/post events, we will have to reorder the groups since one single dataframe is returned. Would it be worth creating a `is_next_pre` column like this to select the rows to keep ? This would allow to conserve the order.

```
new_columns = dict(
        delta=grouped['Time'].transform(lambda time: time.diff().shift(-1)),
        is_next_pre=grouped["is_pre"].shift(-1),
)
```

Reply:
You will still need to detect if the first event of each group is a "post", which can be done with something like that:
```
def f(df):
    df = df.shift(-1) if not df['is_pre'].iat[0] else df
    df['delta'] = df['Time'].diff()
    return df

grouped.apply(f).reset_index('Time').sort_index()
```
This is pretty much the same thing as I mentioned earlier except that pandas is doing the concatenation of each group dataframe on its own. In my experiment, it does mess up the order so it needs to be restored by re-sorting against the time.

This is the point where pandas API falls a little bit apart: `GroupBy.apply()` returns a dataframe rather than another GroupBy object, which means you cannot easily chain operations on it (you have to put all of it in one function and use it with `apply()`). This means that any "advanced" processing *has* to be done with `GroupBy.apply/map/transform()`, unless it can be done in a single step with a `GroupBy` method like `GroupBy.diff()`.

Wrt to broken sequences of pre/post, it probably makes more sense to detect them and just raise a ValueError rather than attempt to silently cover it, as it is probably a sign of grouping on the wrong columns (or bug in the kernel code that emitted the events).

The only case that would really make sense would be sequences like `pre1 -> pre2 -> post2 -> post1` that could occur e.g. in function entry/exit events for a recursive one.

Reply:
I don't understand why the case where the first event of a group is a "post"
event is problematic, we are only keeping the "pre" events with
`     df = df.loc[df["is_pre"]]`
so all the deltas between a "post" and a "pre" event should be tossed in the
end.

About the broken sequences, I agree throwing an error when the sequence is
wrong would be rightful.
Replacing:
```
# For each group, select the rows that are correctly interleaving
# (pre-post-pre-post-...). If multiple rows coming from the same dataframe
# (pre_df or post_df), the first row is kept.
grouped = df.groupby(group_on, observed=True, sort=False)
df = df.loc[grouped["is_pre"].transform(lambda x: x != x.shift(-1)).fillna(False)]
```

by this:
```
# In each group, search for a faulty sequence (where pre/post events are not interleaving,
# e.g. pre1->pre2->post1->post2). In such case, raise an error.
df_faulty = df.loc[grouped["is_pre"].transform(lambda x: x == x.shift(-1)).fillna(False)]
if (not df_faulty.empty):
    raise ValueError
```

should allow to detect this.

Reply:
>  df = df.loc[df["is_pre"]]
Good point, I got confused, I think I was under the wrong assumption that we would filter the results by getting one row over two, which would only work if everything is properly aligned (it would probably be a bit faster but is much less safe).

># In each group, search for a faulty sequence (where pre/post events are not interleaving,
# e.g. pre1->pre2->post1->post2). In such case, raise an error.
df_faulty = df.loc[grouped["is_pre"].transform(lambda x: x == x.shift(-1)).fillna(False)]
if (not df_faulty.empty):
    raise ValueError

You can avoid creating a new dataframe with something like:
```
if grouped['is_pre'].transform(lambda x: x == x.shift()).any():
    raise ValueError('Unexpected sequence of pre and post event (more than one "pre" or "post" in a row')
```